lapis2/src/main/kotlin/org/genspectrum/lapis/controller/CompressionFilter.kt

```diff
@@ -0,0 +1,141 @@
+fun ZstdOutputStream.commitUnderlyingResponseToPreventContentLengthFromBeingSet() {
+    val nothing = ByteArray(0)
+    write(nothing)
+}
+
+@Component
+@Order(DOWNLOAD_AS_FILE_FILTER_ORDER - 1)
+class CompressionFilter(val objectMapper: ObjectMapper) : OncePerRequestFilter() {
+    override fun doFilterInternal(
+        request: HttpServletRequest,
+        response: HttpServletResponse,
+        filterChain: FilterChain,
+    ) {
+        val reReadableRequest = CachedBodyHttpServletRequest(request, objectMapper)
+
+        val requestWithContentEncoding = HeaderModifyingRequestWrapper(
+            reReadableRequest = reReadableRequest,
+            headerName = ACCEPT_ENCODING,
+            computeHeaderValueFromRequest = ::computeAcceptEncodingValueFromRequest,
+        )
+
+        val maybeCompressingResponse = createMaybeCompressingResponse(
+            response,
+            requestWithContentEncoding.getHeaders(ACCEPT_ENCODING),
+        )
+
+        filterChain.doFilter(
+            requestWithContentEncoding,
+            maybeCompressingResponse,
+        )
+
+        maybeCompressingResponse.outputStream.flush()
+        maybeCompressingResponse.outputStream.close()
+    }
+
+    private fun createMaybeCompressingResponse(
+        response: HttpServletResponse,
+        acceptEncodingHeaders: Enumeration<String>?,
+    ) = when (val compression = Compression.fromHeaders(acceptEncodingHeaders)) {
+        null -> response
+        else -> CompressingResponse(response, compression)
+    }
+}
+
+private fun computeAcceptEncodingValueFromRequest(reReadableRequest: CachedBodyHttpServletRequest) =
+    when (reReadableRequest.getStringField(COMPRESSION_PROPERTY)) {
+        Compression.GZIP.value -> Compression.GZIP.value
+        Compression.ZSTD.value -> Compression.ZSTD.value
+        else -> null
+    }
+
+class CompressingResponse(
+    response: HttpServletResponse,
+    compression: Compression,
+) : HttpServletResponse by response {
+    init {
+        log.info { "Compressing using $compression" }
+        response.setHeader(CONTENT_ENCODING, compression.value)
+        preventSpringFromSettingTheContentLengthWhichIsUnknownWhenCompressing(response)
+    }
+
+    private fun preventSpringFromSettingTheContentLengthWhichIsUnknownWhenCompressing(response: HttpServletResponse) {
+        response.addHeader(TRANSFER_ENCODING, "chunked")
+    }
+
+    private val servletOutputStream = CompressingServletOutputStream(response.outputStream, compression)
+
+    override fun getOutputStream() = servletOutputStream
+}
+
+class CompressingServletOutputStream(
+    private val outputStream: ServletOutputStream,
+    compression: Compression,
+) : ServletOutputStream() {
+    private val compressingStream = compression.compressionOutputStreamFactory(outputStream)
+
+    override fun write(b: Int) {
```

Comment:
Is something special, or can we name it "byte" here?

Reply:
Ah no, I forgot to rename it. IntelliJ took this from the function that it overwrites.